filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,150 @@
+package auth
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/MicahParks/keyfunc"
+	"github.com/golang-jwt/jwt"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+)
+
+const (
+	JwtValidationName = "jwtValidation"
+)
+
+type (
+	jwtValidationSpec struct {
+		options TokenintrospectionOptions
+	}
+
+	jwtValidationFilter struct {
+		jwksUri string
+	}
+)
+
+var m sync.Mutex
+
+var refreshInterval = time.Hour
+var refreshRateLimit = time.Minute * 5
+var refreshTimeout = time.Second * 10
+var refreshUnknownKID = true
+
+//the map of jwks keyfunctions stored per jwksUri
+var jwksMap map[string]*keyfunc.JWKs = make(map[string]*keyfunc.JWKs)
+
+func NewJwtValidationWithOptions(o TokenintrospectionOptions) filters.Spec {
+	return &jwtValidationSpec{
+		options: o,
+	}
+}
+
+func (s *jwtValidationSpec) Name() string {
+	return JwtValidationName
+}
+
+func (s *jwtValidationSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	if len(args) != 1 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+
+	issuerURL := sargs[0]
+
+	cfg, err := getOpenIDConfig(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = getKeyFunction(cfg.JwksURI, true)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &jwtValidationFilter{
+		jwksUri: cfg.JwksURI,
+	}
+
+	return f, nil
+}
+
+func getKeyFunction(url string, creationAllowed bool) (jwks *keyfunc.JWKs, err error) {
```

Comment:
I do not get it - why do we need `creationAllowed bool` if it is true in both `CreateFilter` and `Request` call sites.

If the goal is to ensure that keyFunc exists before `Request` then lets have two methods:
- `registerKeyFunc(jwksUri)` that would create the function, and put it into the map
- `getKeyFunc(jwksUri)` - this simply returns a value from the map, no need to check anything as the `jwksUri` is the same as was used  during creation

Reply:
it was a typo, creationAllowed should be false for request. Fixed this in nex commit

Reply:
@abinet The point is that there are two actions `register` and `get` that should be separate functions - the goal is clarity.
E.g. `register` does not have to return keyFunk so
```go
_, err = getKeyFunction(cfg.JwksURI, true)
```
would be
```go
err := registerKeyFunction(cfg.JwksURI)
```
while `get` does not have to check the presence of the value as it always has to be there after `register` so
```go
jwks, err := getKeyFunction(f.jwksUri, false)
if err != nil {
...
}
```
becomes
```go
jwks := getKeyFunction(f.jwksUri)
```